### DISCUSSION TITLE
Mocking a Riverpod-generated AutoDisposeNotifier results in `...has no instance method '_setElement'.`

### DISCUSSION THREAD

**Discussion by derolf** [rrousselGit/riverpod#2514] [Q&A] [answered]
I have a Notifier:

```
@riverpod
class Foo extends _$Foo {
...
}
```

I am mocking it with mocktail:

```
class MockFoo extends Mock implements Foo {}
```

Supplying is at an override yields this error:

```
class 'MockFoo' has no instance method '_setElement'.
Receiver: Instance of 'MockFoo'
Tried calling: _setElement(Instance of 'AutoDisposeNotifierProviderElement<Foo, FooState>')
```

What should I do?

**Comment by rrousselGit** [comment]
extends Foo with Mock

**Reply by derolf** [reply]
> extends Foo with Mock

That exposes all methods from `Foo` which I want to "hide".

**Comment by derolf** [comment, marked as answer]
This works:

```
class MockFoo extends AutoDisposeNotifier<Bar> with Mock implements Foo {
}
```

**Reply by vdmytrovskyi-hpe** [reply]
Holly cow! How can someone come up with this?
Thank you for your answer!

**Comment by Takrem1** [comment]
Hey i have the same Problem, but the solution above does not work for me.

With the Line:

```dart
class MockMyController extends AutoDisposeNotifier<MyModel> with Mock implements MyController {}
```

i get this Error:
```
Superinterfaces don't have a valid override for 'build': AutoDisposeNotifier.build (MyModel Function()), MyController.build (MyModel Function(group)).
```

Both of the classes (AutoDisposeNotifier and MyController) have a build() method. I tried to override both of them, but this was not possible for me. Any Suggestions how to solve this Problem?

This is my Code:
```dart
class MockGroup extends Mock implements Group {}
// next line has the error
class MockMyController extends AutoDisposeNotifier<MyModel> with Mock implements MyController {}

void main() {
  group('widget test group', () {
    testWidgets('test', (tester) async {
      final mockGroup = MockGroup();
      final mockMyController = MockMyController();

      await tester.pumpWidget(
        ProviderScope(
          overrides: [
            // myControllerProvider has the type MyControllerFamily. I get the specific MyControllerProvider with the group-parameter
            myControllerProvider(mockGroup).overrideWith(() => mockMyController),
          ],
          child: MaterialApp(
            home: Scaffold(body: TestWidget),
          ),
        ),
      );
    });
  });
}
```

**Reply by Randomguy01** [reply]
I had the same issue and fixed it by changing `AutoDisposeNotifier<MyModel>` to `AutoDisposeFamilyNotifier<MyModel, arg>` where arg was the type passed to the family modifier.

**Reply by tomdittrich** [reply]
Your tip has brought me a little further. However, I still get an error message when I try to read the provider:

`type ‘(MyArg)’ is not a subtype of type ‘MyArg’ in type cast`

`MyArg` is the type passed to the family modifier.

**Comment by tazik561** [comment]
> extends AutoDisposeNotifier<Bar>

how did you return state ?
